lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2978,17 +2978,24 @@ static bool checkDifferentiationParameters(
   SmallVector<Type, 4> wrtParamTypes;
   indices->getSubsetParameterTypes(functionType, wrtParamTypes);
   for (unsigned i : range(wrtParamTypes.size())) {
+    SourceLoc loc = parsedWrtParams.empty()
+        ? attrLoc
+        : parsedWrtParams[i].getLoc();
     auto wrtParamType = wrtParamTypes[i];
+    if (wrtParamType->is<InOutType>()) {
```

Comment:
I'm not sure if `InOutType` is the right thing to check for (this may be why you are not getting the error). Inout is formally encoded as part of `ParameterTypeFlags`, which is part of `AnyFunctionType::Param`. `AutoDiffParameterIndices::getSubsetParameterTypes` currently takes a list of `Type`s, not `AnyFunctionType::Param`s. I think we can change `AutoDiffParameterIndices::getSubsetParameterTypes` to something else, roughly:
```cpp
void getSubsetParameters(AnyFunctionType *type, SmallVectorImpl<AnyFunctionType::Param> &result) const;
```

This will return `AnyFunctionType::Param`s instead of just the `Type`s. Then you can check whether a parameter is `inout` by looking at the parameter flags.

Reply:
The issue with this is that I can't forward declare a nested class (i.e., `AnyFunctionType::Param`) and so I can't declare such a function in `AutoDiff.h` without other changes. What do you suggest? I have an implementation currently with the only error being that related to forward declaration.

Reply:
There is a hacky solution: Use `AnyFunctionType::Param::getOldType()` in `AutoDiffParameterIndices::getSubsetParameterTypes`.  This method returns an `InOutType` when the parameter is `inout`. I think this will be deprecated, but we can use it for now.

Reply:
I can either do that or properly implement the check. I'm fine either way but for the latter we should probably discuss a nice way to do it that avoids the forward declaration issue. What if I pull `Param` out of `AnyFunctionType`? Would that be acceptable if it doesn't require too many changes?

Reply:
Pulling it out won't be acceptable because we should try to minimize changes to the existing Swift infrastructure. Using `getOldType()` is fine for now.